index.js

```diff
@@ -36,11 +37,40 @@ var reservedRe = /^\.+$/;
 var windowsReservedRe = /^(con|prn|aux|nul|com[0-9]|lpt[0-9])(\..*)?$/i;
 var windowsTrailingRe = /[\. ]+$/;
 
-function sanitize(input, replacement) {
+function containsInvalids(input, invalids) {
+  for (var i = 0; i < invalids.length; i++) {
+    if (input.indexOf(invalids[i]) !== -1) {
+      return true;
+    }
+  }
+  return false;
+}
+
+function sanitize(input, replacement, invalids) {
   if (typeof input !== 'string') {
     throw new Error('Input must be string');
   }
-  var sanitized = input
+  if (typeof replacement === 'string' && invalids.indexOf(replacement) !== -1) {
+    throw new Error("The replacement string can't be part of options.additionalInvalidStrings or contain substrings which are part of options.additionalInvalidStrings!");
+  }
+  var sanitized = input;
+  var counter = 0;
+  while (containsInvalids(sanitized, invalids)) {
+    for (var i = 0; i < invalids.length; i++) {
+      if (typeof replacement !== 'string') {
+        var tempReplacement = replacement(invalids[i]);
```

Comment:
To improve performance, it would be nice to cache these replacement values outside the loops. Before starting any of the loops, you could create a map of `{ invalid: replacement }` and then use that here instead of calling the function for each iteration.

Something like:

```js
var replacements;
if (typeof replacement === 'function') {
  replacements = invalidStrings.reduce(function (replacements, invalidString) {
    var replacementString = replacement(invalidString);
    // you could sanitize the replacement string here to avoid the loop
    replacements[invalidString] = replacementString;
    return replacements;
  }, {});
}
// then the for-loop:
for (...) {
}
```
Remember, that function will return the same value for each different invalid string, therefore there's no need calling it again each time we encounter the same invalid string.

Reply:
That is a good idea which I didn't think of, at least at first.

Be careful though, you don't know the nature of the replacement function!  
What if it uses some kind of counter mechanism or a timestamp with milliseconds for returning the replacement?  
Just because we're only passing the string to the function, doesn't mean that this is the only thing it depends on ;)

Reply:
Haha true, one can never anticipate userland creativity. I think we wouldn't be faulted for the optimisation but it's fine by me to leave the current implementation as is.